### ISSUE TITLE
Segwit Transaction Digest Algorithm is broken

### ISSUE THREAD

**Issue by acrespo** [bitcoinj/bitcoinj#1935]
The current implementation of `Transaction#hashForWitnessSignature` does not exactly  follow the BIP143 [specification](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification). It's missing one tiny bit of information (the scriptCode length).

I've built a small test using BIP143's test examples data:

 ```
public class SegwitDigestTest {

    private static final TestNet3Params TESTNET = TestNet3Params.get();

    @Test
    public void testTransactionDigestAndSignatureForP2sh_P2wsh() {

        final String rawhex = "010000000136641869ca081e70f394c6948e8af409e18b619df2ed74aa106c1ca297"
                + "87b96e0100000000ffffffff0200e9a435000000001976a914389ffce9cd9ae88dcc0631e88a821f"
                + "fdbe9bfe2688acc0832f05000000001976a9147480a33f950689af511e6e84c138dbbd3c3ee41588"
                + "ac00000000";

        final Transaction transaction = new Transaction(TESTNET, Utils.HEX.decode(rawhex));

        // Values taken from Bip 143 example for P2SH-P2WSH
        // See https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#p2sh-p2wsh
        final String witnessScriptHex = "56210307b8ae49ac90a048e9b53357a2354b3334e9c8bee813ecb98e99"
                + "a7e07e8c3ba32103b28f0c28bfab54554ae8c658ac5c3e0ce6e79ad336331f78c428dd43eea8449b"
                + "21034b8113d703413d57761b8b9781957b8c0ac1dfe69f492580ca4195f50376ba4a21033400f6af"
                + "ecb833092a9a21cfdf1ed1376e58c5d1f47de74683123987e967a8f42103a6d48b1131e94ba04d97"
                + "37d61acdaa1322008af9602b3b14862c07a1789aac162102d8b661b0b3302ee2f162b09e07a55ad5"
                + "dfbe673a9f01d9f0c19617681024306b56ae";

        final int inputIndex = 0;
        final int valueInSatoshis = 987654321;
        final Transaction.SigHash sigHashType = Transaction.SigHash.ALL;

        final Sha256Hash sha256Hash = transaction.hashForWitnessSignature(
                inputIndex,
                Utils.HEX.decode(witnessScriptHex),
                Coin.valueOf(valueInSatoshis),
                sigHashType,
                false
        );

        final String expected = "185c0be5263dce5b4bb50a047973c1b6272bfbd0103a89444597dc40b248ee7c";
        assertEquals(expected, sha256Hash.toString());
    }
}
```

which the current implementation fails to pass.

I've a working fork and I can open a PR if you like.
You can see the fix in [this commit](https://github.com/acrespo/bitcoinj/commit/9778a7df8b69d371868b52ecc123fc2b1cdb014e)

**Comment by schildbach** [comment]
Thanks for looking at this! This is weird. By looking at the spec I tend to agree with you, but in reality transactions spending a P2WPKH output have been accepted by the network.

I also applied your patch to master and it fails two existing tests:

```
org.bitcoinj.core.TransactionTest > testWitnessSignatureP2WPKH FAILED
    org.junit.ComparisonFailure: expected:<[c37af31116d1b27caf68aae9e3ac82f1477929014d5b917657d0eb49478cb670]> but was:<[7e58a8552e119dbf5150469fb8900aa29f18962d89b481fe36a7395c211432f8]>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at org.bitcoinj.core.TransactionTest.testWitnessSignatureP2WPKH(TransactionTest.java:247)

org.bitcoinj.core.TransactionTest > testWitnessSignatureP2SH_P2WPKH FAILED
    org.junit.ComparisonFailure: expected:<[64f3b0f4dd2bb3aa1ce8566d220cc74dda9df97d8490cc81d89d735c92e59fb6]> but was:<[9f89fd22be23214bddc84b48a0d5f081d239b12fb7093005ffbf16df8f078dc4]>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at org.bitcoinj.core.TransactionTest.testWitnessSignatureP2SH_P2WPKH(TransactionTest.java:323)
```

I must admit I didn't test P2SH-P2WPKH and P2SH-P2WSH much, except for the existing unit tests. But P2WPKH is very well tested nowadays, since it's rolled out to millions of wallets since spring. I've never seen a case of a bad sig.

A PR against master is welcome! If it is merged I will backport it to older release branches as needed.

**Comment by schildbach** [comment]
Honestly I think the test vectors in BIP-143 are pretty broken in general. There is [another mistake](https://github.com/bitcoin/bips/pull/748) in https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#native-p2wpkh, we reported it almost a year ago and nobody seems to care. I wouldn't be surprised if the P2SH_P2WSH vectors stem from some early state of the SegWit spec.

**Comment by acrespo** [comment]
Ugh. That's weird about the `Native P2WPKH` test vectors. Didn't know about that. 

I think the `P2SH_P2WSH` are correct since we at Muun are having transactions spending `P2SH_P2WSH` outputs accepted by the network.

Maybe our misunderstandings may be due to the fact that the spec treats `P2WPKH` and `P2WSH` differently:

> The item 5:
> 
> For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac.
> For P2WSH witness program,
> - if the witnessScript does not contain any OP_CODESEPARATOR, the scriptCode is the witnessScript serialized as scripts inside CTxOut.
> - if the witnessScript contains any OP_CODESEPARATOR, the scriptCode is the witnessScript but removing everything up to and including the last executed OP_CODESEPARATOR before the signature checking opcode being executed, serialized as scripts inside CTxOut. (The exact semantics is demonstrated in the examples below)

Though I think that ultimately they are the same as `0x19` is the op code for a `PUSH25` while the script for `P2WPKH` is 25 bytes long. Which is "kinda" what my patch is about but for `P2SH-P2WSH`.

I believe that the difference between the test I built and the two failing tests relies on the way we call `hashForWitnessSignature`. Those tests (as well as other methods like `Transaction#addSignedInput` and `LocalTransactionSigner#signInputs`) call the `hashForWitnessSignature` with the `scriptCode` param having the initial `OP_PUSH25` attached, while my patch does that work **inside** `hashForWitnessSignature`, that's actually the single line added:  

```
(...)
            bos.write(hashPrevouts);
            bos.write(hashSequence);
            bos.write(inputs.get(inputIndex).getOutpoint().getHash().getReversedBytes());
            uint32ToByteStreamLE(inputs.get(inputIndex).getOutpoint().getIndex(), bos);
            bos.write(new VarInt(scriptCode.length).encode());      // THIS IS THE ADDED LINE
            bos.write(scriptCode);
            uint64ToByteStreamLE(BigInteger.valueOf(prevValue.getValue()), bos);
            uint32ToByteStreamLE(inputs.get(inputIndex).getSequenceNumber(), bos);
            bos.write(hashOutputs);
            uint32ToByteStreamLE(this.lockTime, bos);
            uint32ToByteStreamLE(0x000000ff & sigHashType, bos);
        } catch (IOException e) {
            throw new RuntimeException(e);  // Cannot happen.
        }

        return Sha256Hash.twiceOf(bos.toByteArray());
    }
```

Another difference (and I think this is more important), is that while my patch uses a `VarInt`, the current code uses a `PUSHDATA`, which may work identically for scripts up to 75 bytes, but not for larger scripts. For `P2WPKH` the script length is fixed and its well below that threshold (24 bytes) so there's no problem there, but since `P2WSH` can allow for larger scripts I think that `VarInt` should be used.   

In other words, yes, the current implementation works fine for `P2WPKH` and `P2SH_P2WPKH`, my patch breaks those scenarios since for those cases the scriptCode length would appear twice, but I think that we should move that behavior inside of `hashForWitnessSignature` and use a `VarInt` for several reasons:

- to allow callers to simply pass their scripts (without making them prefix the script length)
- to handle `P2WSH` correctly

**Comment by schildbach** [comment]
> I think that we should move that behavior inside of `hashForWitnessSignature` and use a `VarInt` to:
> 
>     * allow callers to simply pass their scripts (without making them prefix the script length)

For the P2WPKH case, I disagree. Imho the spec makes is clear that the `0x19` is part of the `scriptCode`.

Also, `VarInt` is different to a `PUSHDATA` operator, so I'm skeptical that this would be the right choice for inside a script. Have you looked at `ScriptBuilder.data()`?

Last but not least, it would be a breaking chance to the `hashForWitnessSignature` and `calculateWitnessSignature` APIs with hard to detect failure modes (unless the change the method signatures).

>     * handle `P2WSH` correctly

I agree.

If I read the spec correctly, there is a `scriptCode` which is prefixed by the `PUSHDATA` op. And a `witnessScript` which isn't.

I'm currently not sure where we should put the adaption from `witnessScript` to `scriptCode`. I think I'd look at how P2SH is currently implemented, there must be some adaption there too. Also it might be worth to peek at the Bitcoin Core source code or Electrum.
